nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2023 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_route/breadth_first_search.hpp"
+
+#include <iostream>
+
+namespace nav2_route
+{
+
+void BreadthFirstSearch::initMotionModel(int x_size, int y_size)
+{
+  max_index_ = x_size * y_size;
+  x_size_ = x_size;
+
+  neighbors_grid_offsets = {-1, +1,
+                            -x_size, +x_size,
+                            -x_size - 1, -x_size + 1,
+                            +x_size - 1, +x_size + 1};
+
+  std::cout << "offset size " << neighbors_grid_offsets.size() << std::endl;
+}
+
+void BreadthFirstSearch::setCostmap(nav2_costmap_2d::Costmap2D *costmap)
+{
+  std::cout << "Set costmap " << std::endl;
+  costmap_ = costmap;
+
+  unsigned int x_size = costmap_->getSizeInCellsX();
+  unsigned int y_size = costmap_->getSizeInCellsY();
+
+  initMotionModel(static_cast<int>(x_size), static_cast<int>(y_size));
+}
+
+void BreadthFirstSearch::setStart(unsigned int mx, unsigned int my) {
+  start_ = getIndex(mx, my, x_size_);
+  states_.insert(std::make_pair(start_, State()));
+}
+
+void BreadthFirstSearch::setGoals(std::vector<unsigned int> mxs, std::vector<unsigned int> mys)
+{
+  goals_.clear();
+  for(unsigned int i = 0; i < (mxs.size() && mys.size()); ++i) {
+    unsigned int index = getIndex(mxs[i], mys[i], x_size_);
+    goals_.push_back(index);
+    states_.insert(std::make_pair(index, State()));
+  }
+}
+
+bool BreadthFirstSearch::search(Coordinates & closest_goal)
+{
+  std::cout << "Start " << std::endl;
+  states_.clear();
+
+  // clear the queue
+  std::queue<unsigned int> empty_queue;
+  queue_.swap(empty_queue);
+
+  //Add start to queue
+  queue_.push(start_);
+
+  std::cout << "Added start " << std::endl;
+
+  while (!queue_.empty()) {
+
+    std::cout << "Queue Size: " << queue_.size() << std::endl;
+
+    unsigned int current = queue_.front();
+    std::cout << "grab front" << std::endl;
+
+    Coordinates current_coord = getCoords(current);
+    std::cout << "Current Node: " << current_coord.x << " " << current_coord.y << std::endl;
+
+    queue_.pop();
+    states_[current].visited = true;
```

Comment:
Woof, these are each individually lookups. These are pricy. Definitely only do this lookup once `auto & curr_state = states_[current];` and use that reference so you don't double that overhead. 

Ideally, we should try to make this something that doesn't involve a lookup. If you store in the `queue_` a class containing a pointer to the node instead of the index, then you can insert an expanded node into the graph that is an unordered map, but then never have to query for it again.

If you look at the smac planner, that's basically what I do. I have a holding class containing the node pointer. So when we get the node from the queue, we have the memory to ask `node->isQueued()` without having to ever look up the data structure which contains the graph. While unordered maps are O(1), constant time doesn't mean _fast_ time when compared to direct memory access